pkg/kube/controllers/queue.go

```diff
@@ -0,0 +1,168 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"go.uber.org/atomic"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/util/workqueue"
+)
+
+type Enqueuer interface {
+	Add(item interface{})
+}
+
+// Queue defines an abstraction around Kubernetes' workqueue.
+// Items enqueued are deduplicated; this generally means relying on ordering of events in the queue is not feasible.
+type Queue struct {
+	queue       workqueue.RateLimitingInterface
+	initialSync *atomic.Bool
+	name        string
+	maxAttempts int
+	work        func(key interface{}) error
+}
+
+// WithName sets a name for the queue. This is used for logging
+func WithName(name string) func(q *Queue) {
+	return func(q *Queue) {
+		q.name = name
+	}
+}
+
+// WithRateLimiter allows defining a custom rate limitter for the queue
+func WithRateLimiter(r workqueue.RateLimiter) func(q *Queue) {
+	return func(q *Queue) {
+		q.queue = workqueue.NewRateLimitingQueue(r)
+	}
+}
+
+// WithMaxAttempts allows defining a custom max attempts for the queue. If not set, items will be discarded after 5 attempts
+func WithMaxAttempts(n int) func(q *Queue) {
+	return func(q *Queue) {
+		q.maxAttempts = n
+	}
+}
+
+// WithReconciler defines the to handle items on the queue
+func WithReconciler(f func(name types.NamespacedName) error) func(q *Queue) {
+	return func(q *Queue) {
+		q.work = func(key interface{}) error {
+			return f(key.(types.NamespacedName))
```

Comment:
Is it better to type assert here and return an error? 
If there is no way to enqueue a `work` func that doesn't use `NamespacedName`, should we even export `Add(item interface{}`?

Reply:
I was going to have a function that allowed `interface{}` but turned out that nothing needed it (yet??). Likely its a bad idea to use anything else, so maybe we should just clean this up and not export Add

Reply:
why not the args consistent?

Reply:
I don't know what you mean by args consistent?

Reply:
interface{} vs types.NamespacedName

Reply:
NamespacedName is the external interface. Internally we pass sync signal as well so it's generic